fragment/fragment/src/main/java/androidx/fragment/app/strictmode/FragmentStrictMode.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.fragment.app.strictmode;
+
+import android.annotation.SuppressLint;
+import android.os.Process;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.RestrictTo;
+import androidx.annotation.VisibleForTesting;
+import androidx.fragment.app.Fragment;
+import androidx.fragment.app.FragmentManager;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * FragmentStrictMode is a tool which detects things you might be doing by accident and brings
+ * them to your attention so you can fix them. Basically, it's a version of
+ * {@link android.os.StrictMode} specifically for fragment-related issues.
+ *
+ * <p>You can decide what should happen when a violation is detected. For example, using {@link
+ * Policy.Builder#penaltyLog} you can watch the output of <code>adb logcat</code> while you
+ * use your application to see the violations as they happen.
+ */
+@SuppressLint("SyntheticAccessor")
+@RestrictTo(RestrictTo.Scope.LIBRARY) // TODO: Make API public as soon as we have a few checks
+public final class FragmentStrictMode {
+    private static final String TAG = "FragmentStrictMode";
+    private static Policy defaultPolicy = Policy.LAX;
+
+    private enum Flag {
```

Comment:
If we're concerned about performance, this could also be a bitmask instead

Reply:
No, the Flags are find and make this clear.